### DISCUSSION TITLE
Make `GenSym` singleton object

### DISCUSSION THREAD

**Discussion by jaschdoc** [flix/flix#8587] [General]
Currently, `GenSym` lives in the `Flix` instance but we could refactor it into its own singleton object. Then if you need to generate a fresh symbol in some phase, it avoids polluting the entire phase with `implicit flix: Flix` everywhere. This is what happened in https://github.com/flix/flix/pull/8586

Also, the `Flix` instance both generates symbols but also allows any phase to call compilation recursively which seems dangerous. Of course, we check that this doesn't happen through tests and PR reviews, but it still seems that compilation and symbol generation could be split into separate interfaces.

To summarize:
1. I suggest `GenSym` should be a single global instance
2. Investigating what other things could be moved to separate interfaces (note that `Flix` can just implement those) to avoid phases being able to call `compile` etc.

However, (2) is not really important or helpful as it easy to avoid and adding more types just increases complexity but I still want to mention it as it *feels wrong* to have the possibility for bad behavior.
